src/Microsoft.DotNet.ServiceFabric.ServiceHost/DelegatedService.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.ServiceFabric.ServiceHost
+{
+    public static class DelegatedService
+    {
+        public static async Task RunServiceLoops<T>(IServiceProvider services, CancellationToken cancellationToken, params Func<CancellationToken, Task>[] loops)
+        {
+            using var linkedToken = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            await Lifecycle.OnStartingAsync(services);
+            var logger = services.GetRequiredService<ILogger<T>>();
+            List<Task> pendingTasks = new List<Task>();
+            try
+            {
+                await using var _ =
+                    cancellationToken.Register(() => logger.LogInformation("Service abort cancellation requested"));
+                logger.LogInformation("Entering service 'RunAsync'");
+                pendingTasks = loops.Select(l => l(linkedToken.Token)).ToList();
+                var finished = await Task.WhenAny(pendingTasks);
+                pendingTasks.Remove(finished);
+                await finished;
+                logger.LogWarning("Abnormal service exit without cancellation");
+            }
+            catch (OperationCanceledException e) when (e.CancellationToken == linkedToken.Token)
```

Comment:
Should this be linkedToken.Token, or just cancellationToken? This block is for "SF canceled us" right?

Reply:
I assume that "cancellationToken" getting cancelled with ALSO cancel the linked one. Since the linked one is what we gave to all the callbacks, that _should_ be the one present in the exception.

Reply:
Yup. Which experiment shows that if you have `b = CreateLinkedTokenSource(a)`, then even if a is cancelled, b.Token is the thing present in the exception.